### PR TITLE
BUG: DataFrame.shift with axis=1 and mismatched fill_value

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/internals/managers.py [hunk 1]
```diff
@@ -362,7 +363,22 @@ def shift(self: T, periods: int, axis: int, fill_value) -> T:
         if fill_value is lib.no_default:
             fill_value = None
 
-        if axis == 0 and self.ndim == 2 and self.nblocks > 1:
+        if (
+            axis == 0
```
Comment (jreback) [inline pandas/core/internals/managers.py:367]: this is super complicated can you split into multiple clauses & add comments

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
user facing?

File: pandas/core/internals/managers.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/managers.py:367]: sure will comment.  fortunately we can rip out some of this in 2.0

**Comment by jbrockmendel** [comment]
> user facing?

yes, there's a whatsnew (forgot to check the checkbox until just now)

File: pandas/core/internals/managers.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/managers.py:367]: comment added + green

**Review by jreback — APPROVED** [review]

**Pull request merged**